### PR TITLE
support different replica count per node.

### PR DESCRIPTION
**Pull request by jhgg**
This PR adds support for allowing different replica counts per node to `HashRing.ETS`.

### CODE REVIEW

File: lib/utils.ex [hunk 1]
```diff
@@ -8,17 +8,21 @@ defmodule HashRing.Utils do
...
   end
 
-  defp gen_items([], _replicas, items) do
+  defp do_gen_items([], items) do
     Enum.sort(items, &(elem(&1, 0) < elem(&2, 0)))
   end
-  defp gen_items([node|nodes], replicas, items) do
-    items = Enum.reduce(replicas, items, &([{hash("#{node}#{&1}"), node}|&2]))
-    gen_items(nodes, replicas, items)
+  defp do_gen_items([{node, num_replicas}|nodes], items) do
```
Comment (scohen) [inline lib/utils.ex]: shadowing the node function here, space around the pipe.

File: lib/utils.ex [hunk 2]
```diff
@@ -8,17 +8,21 @@ defmodule HashRing.Utils do
...
+    items = Enum.reduce(0..(num_replicas - 1), items, &([{hash("#{node}#{&1}"), node}|&2]))
```
Comment (scohen) [inline lib/utils.ex]: is it nit-picky if I want spaces around the pipe here? 
also, the captured function looks like swearing to me. Perhaps: 
```elixir
fn replica, acc -> 
 replica_name = hash("#{node}#{replica}")
 [{replica_name, node}, | acc] 
end
```

File: lib/utils.ex [hunk 3]
```diff
@@ -8,17 +8,21 @@ defmodule HashRing.Utils do
...
+  @spec gen_items([{binary, integer}]) :: [{integer, binary}]
   @spec gen_items(binary, integer) :: [{integer, binary}]
+  def gen_items([]), do: []
+  def gen_items(nodes), do: do_gen_items(nodes, [])
   def gen_items([], _num_replicas), do: []
-  def gen_items(nodes, num_replicas) do
-    gen_items(nodes, Enum.to_list(0..(num_replicas - 1)), [])
+  def gen_items(nodes, default_num_replicas) do
+    nodes = for node <- nodes, do: {node, default_num_replicas}
+    gen_items(nodes)
```
Comment (scohen) [inline lib/utils.ex]: call do_gen_items here?

File: lib/hash_ring/ets.ex [hunk 4]
```diff
@@ -123,26 +128,34 @@ defmodule HashRing.ETS do
...
 
-  def handle_call({:set_nodes, nodes}, _from, state) do
+  def handle_call({:set_nodes, nodes}, _from, %{default_num_replicas: default_num_replicas}=state) do
+    nodes = transform_nodes(nodes, default_num_replicas)
     {:reply, {:ok, nodes}, rebuild(%{state | nodes: nodes})}
   end
-  def handle_call({:add_node, node_name}, _from, %{nodes: nodes}=state) do
-    if node_name in nodes do
+  def handle_call({:add_node, node_name, nil}, from, %{default_num_replicas: default_num_replicas}=state) do
+    handle_call({:add_node, node_name, default_num_replicas}, from, state)
```
Comment (scohen) [inline lib/hash_ring/ets.ex]: I don't know how I feel about this, calling a callback from a callback.

File: lib/hash_ring/ets.ex [hunk 5]
```diff
@@ -220,4 +232,15 @@ defmodule HashRing.ETS do
       _ -> nil
     end
   end
+
+  defp has_node_with_name?(nodes, node_name) do
+    Enum.any?(nodes, fn {existing_node, _} -> existing_node == node_name end)
```
Comment (scohen) [inline lib/hash_ring/ets.ex]: I thought you'd do something cool like:
```elixir
Enum.any(nodes, &match?(^existing_node, &1))
```

File: lib/utils.ex (hunk 3 above)
Comment (jhgg) [reply lib/utils.ex]: eh, the thought is this function transforms the input into a value acceptable for `gen_items/1`

**Pull request merged**